File: src/codeheart_operating_kit/commands/inspect.py
from __future__ import annotations

import json
from pathlib import Path


TECHNICAL_MARKERS = {
    ".git",
    "pyproject.toml",
    "package.json",
    "Cargo.toml",
    "go.mod",
    "Makefile",
    "AGENTS.md",
    "src",
    "tests",
}


def inspect_folder(path: Path) -> dict[str, object]:
    path = path.expanduser()
    if path.exists() and not path.is_dir():
        return {"path": str(path), "mode": "ambiguous-folder-stop", "reason": "target is not a folder"}
    if not path.exists():
        return {"path": str(path), "mode": "new-folder-setup", "reason": "folder does not exist"}
    entries = {entry.name for entry in path.iterdir()}
    if ".codeheart" in entries and ((path / ".codeheart/kit.lock.yaml").exists() or (path / ".codeheart/kit").exists()):
        return {"path": str(path), "mode": "existing-operating-kit-repair", "reason": "Operating Kit metadata exists"}
    if not entries:
        return {"path": str(path), "mode": "new-folder-setup", "reason": "folder is empty"}
    if entries & TECHNICAL_MARKERS:
        return {
            "path": str(path),
            "mode": "existing-technical-project-adoption",
            "reason": "technical project markers found",
            "markers": sorted(entries & TECHNICAL_MARKERS),
        }
    if len(entries) > 100:
        return {"path": str(path), "mode": "ambiguous-folder-stop", "reason": "folder has many existing files"}
    return {"path": str(path), "mode": "existing-folder-setup", "reason": "folder contains existing files"}


def run(args) -> int:
    result = inspect_folder(Path(args.path))
    if args.json:
        print(json.dumps(result, indent=2, sort_keys=True))
    else:
        print(f"{result['mode']}: {result['reason']}")
    return 0

File: internal/commands/inspect.go
package commands

import (
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
)

var technicalMarkers = map[string]bool{
	".git":           true,
	"pyproject.toml": true,
	"package.json":   true,
	"Cargo.toml":     true,
	"go.mod":         true,
	"Makefile":       true,
	"AGENTS.md":      true,
	"src":            true,
	"tests":          true,
}

func RunInspect(args []string, stdout io.Writer, stderr io.Writer) int {
	values, bools, positionals, err := parseValueArgs(args, map[string]bool{"--json": false})
	_ = values
	if err != nil {
		return writeArgError(stderr, "inspect", err)
	}
	path := "."
	if len(positionals) > 0 {
		path = positionals[0]
	}
	if len(positionals) > 1 {
		return writeArgError(stderr, "inspect", fmt.Errorf("unexpected argument %q", positionals[1]))
	}
	result := InspectFolder(path)
	if bools["--json"] {
		if err := writeJSON(stdout, result); err != nil {
			fmt.Fprintf(stderr, "codeheart-operating-kit inspect: error: %v\n", err)
			return 1
		}
		return 0
	}
	fmt.Fprintf(stdout, "%s: %s\n", result["mode"], result["reason"])
	return 0
}

func InspectFolder(input string) map[string]any {
	path := expandPath(input)
	info, err := os.Stat(path)
	if err == nil && !info.IsDir() {
		return map[string]any{"path": path, "mode": "ambiguous-folder-stop", "reason": "target is not a folder"}
	}
	if os.IsNotExist(err) {
		return map[string]any{"path": path, "mode": "new-folder-setup", "reason": "folder does not exist"}
	}
	entriesList, err := os.ReadDir(path)
	if err != nil {
		return map[string]any{"path": path, "mode": "ambiguous-folder-stop", "reason": err.Error()}
	}
	entries := map[string]bool{}
	for _, entry := range entriesList {
		entries[entry.Name()] = true
	}
	if entries[".codeheart"] && (exists(filepath.Join(path, ".codeheart", "kit.lock.yaml")) || exists(filepath.Join(path, ".codeheart", "kit"))) {
		return map[string]any{"path": path, "mode": "existing-operating-kit-repair", "reason": "Operating Kit metadata exists"}
	}
	if len(entries) == 0 {
		return map[string]any{"path": path, "mode": "new-folder-setup", "reason": "folder is empty"}
	}
	markers := []string{}
	for entry := range entries {
		if technicalMarkers[entry] {
			markers = append(markers, entry)
		}
	}
	if len(markers) > 0 {
		sort.Strings(markers)
		return map[string]any{
			"path":    path,
			"mode":    "existing-technical-project-adoption",
			"reason":  "technical project markers found",
			"markers": stringsAsAny(markers),
		}
	}
	if len(entries) > 100 {
		return map[string]any{"path": path, "mode": "ambiguous-folder-stop", "reason": "folder has many existing files"}
	}
	return map[string]any{"path": path, "mode": "existing-folder-setup", "reason": "folder contains existing files"}
}

func exists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

func stringsAsAny(values []string) []any {
	result := make([]any, len(values))
	for index, value := range values {
		result[index] = value
	}
	return result
}
